// @dimpu/ngx-quill/esm2015/lib/ngx-quill.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, forwardRef, Input, Output, ViewEncapsulation } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/** @type {?} */
let Quill;
export class NgxQuillComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.defaultModules = {
            toolbar: [
                ['bold', 'italic', 'underline', 'strike'],
                ['blockquote', 'code-block'],
                [{ 'header': 1 }, { 'header': 2 }],
                [{ 'list': 'ordered' }, { 'list': 'bullet' }],
                [{ 'script': 'sub' }, { 'script': 'super' }],
                [{ 'indent': '-1' }, { 'indent': '+1' }],
                [{ 'direction': 'rtl' }],
                [{ 'size': ['small', false, 'large', 'huge'] }],
                [{ 'header': [1, 2, 3, 4, 5, 6, false] }],
                [{ 'color': new Array() }, { 'background': new Array() }],
                [{ 'font': new Array() }],
                [{ 'align': new Array() }],
                ['clean'],
                ['link', 'image', 'video'] // link and image, video
            ]
        };
        this.options = {};
        this.blur = new EventEmitter();
        this.focus = new EventEmitter();
        this.ready = new EventEmitter();
        this.change = new EventEmitter();
        this.onModelChange = () => { };
        this.onModelTouched = () => { };
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        /** @type {?} */
        const toolbarElem = this.elementRef.nativeElement.querySelector('[ngx-quill-toolbar]');
        this.elementRef.nativeElement.insertAdjacentHTML('beforeend', '<div ngx-quill-element></div>');
        this.editorElem = this.elementRef.nativeElement.querySelector('[ngx-quill-element]');
        if (!Quill) {
            Quill = require('quill');
        }
        this.options = Object.assign({
            modules: this.defaultModules,
            placeholder: 'Insert text here ...',
            readOnly: false,
            theme: 'snow',
            boundary: document.body
        }, this.options);
        if (toolbarElem) {
            this.options.modules['toolbar'] = toolbarElem;
        }
        this.quillEditor = new Quill(this.editorElem, this.options);
        if (this.content) {
            this.quillEditor.pasteHTML(this.content);
        }
        this.ready.emit(this.quillEditor);
        this.quillEditor.on('selection-change', (range) => {
            if (!range) {
                this.onModelTouched();
                this.blur.emit(this.quillEditor);
            }
            else {
                this.focus.emit(this.quillEditor);
            }
        });
        this.quillEditor.on('text-change', (delta, oldDelta, source) => {
            /** @type {?} */
            let html = this.editorElem.children[0].innerHTML;
            /** @type {?} */
            const text = this.quillEditor.getText();
            if (html === '<p><br></p>') {
                html = null;
            }
            this.onModelChange(html);
            this.change.emit({
                editor: this.quillEditor,
                html: html,
                text: text
            });
        });
        this.quillEditor.enable(false);
        setTimeout(() => {
            this.quillEditor.enable(true);
        }, 0);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['readOnly'] && this.quillEditor) {
            this.quillEditor.enable(!changes['readOnly'].currentValue);
        }
        if (changes['content'] && this.quillEditor) {
            this.quillEditor.pasteHTML(this.content);
        }
    }
    /**
     * @param {?} currentValue
     * @return {?}
     */
    writeValue(currentValue) {
        this.content = currentValue;
        if (this.quillEditor) {
            if (currentValue) {
                this.quillEditor.pasteHTML(currentValue);
                return;
            }
            this.quillEditor.setText('');
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onModelChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onModelTouched = fn;
    }
}
NgxQuillComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-quill',
                template: `<ng-content select="[ngx-quill-toolbar]"></ng-content>`,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => NgxQuillComponent),
                        multi: true
                    }],
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
NgxQuillComponent.ctorParameters = () => [
    { type: ElementRef }
];
NgxQuillComponent.propDecorators = {
    options: [{ type: Input }],
    content: [{ type: Input }],
    blur: [{ type: Output }],
    focus: [{ type: Output }],
    ready: [{ type: Output }],
    change: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    NgxQuillComponent.prototype.quillEditor;
    /** @type {?} */
    NgxQuillComponent.prototype.editorElem;
    /** @type {?} */
    NgxQuillComponent.prototype.defaultModules;
    /** @type {?} */
    NgxQuillComponent.prototype.options;
    /** @type {?} */
    NgxQuillComponent.prototype.content;
    /** @type {?} */
    NgxQuillComponent.prototype.blur;
    /** @type {?} */
    NgxQuillComponent.prototype.focus;
    /** @type {?} */
    NgxQuillComponent.prototype.ready;
    /** @type {?} */
    NgxQuillComponent.prototype.change;
    /** @type {?} */
    NgxQuillComponent.prototype.onModelChange;
    /** @type {?} */
    NgxQuillComponent.prototype.onModelTouched;
    /** @type {?} */
    NgxQuillComponent.prototype.elementRef;
}
